### PR TITLE
chore(archive): archive and drivers refactor

### PR DESCRIPTION
**Pull request by SionoiS**
This is a feature branch used to coordinate a big refactor of archive and all drivers to clean/optimize specifically for Store v3.

This work is required before we start measuring performance of Store v3.

@gabrielmer & @Ivansete-status  already reviewed all the PRs but if you want to double check go ahead.

What's new are non-PR related commits.
 
Related PRs;
- #2751
- #2752 
- #2753 
- #2754 
- #2755 
- #2756 

Closes:
- #2745
- #2637

### CODE REVIEW

**Comment by SionoiS** [comment]
@SionoiS https://github.com/waku-org/nwaku/pull/2756#discussion_r1627156202

**Comment by SionoiS** [comment]
I have some issues to fix before review.

**Comment by SionoiS** [comment]
CI is red but I can't reproduce anything...

**Review by Ivansete-status — COMMENTED** [review]
Thanks for this PR! This is a very complex task and great work so far!
I'm adding some comments that I hope you find useful.
Let's merge when we revisit all the comments and all CI are green (but the js-waku ones)

File: vendor/negentropy
Comment (SionoiS) [reply vendor/negentropy:1]: When I switch branch this sneaky files get added without me noticing, sorry about that.

File: migrations/message_store_postgres/content_script_version_6.nim
Comment (SionoiS) [reply migrations/message_store_postgres/content_script_version_6.nim:3]: You mean all changes have to be done in-place? Can this be done even for a primary key?

File: migrations/message_store_postgres/content_script_version_6.nim
Comment (SionoiS) [reply migrations/message_store_postgres/content_script_version_6.nim:15]: I was thinking that we keep the old values for now so that we are still compatible with v2. Would that not work?

File: tests/all_tests_waku.nim
Comment (SionoiS) [reply tests/all_tests_waku.nim:29]: In this case I'm just adding back the legacy tests.

File: tests/waku_archive/test_driver_queue_index.nim
Comment (SionoiS) [reply tests/waku_archive/test_driver_queue_index.nim:24]: So that we can test equality.

**Comment by SionoiS** [comment]
Latest error in CI
```
ERR 2024-06-12 19:30:49.752+00:00 error creating ArchiveDriver               tid=8120 file=postgres_legacy.nim:14 error="postgres health check error: connRes.isErr in query: pool is not live"
```

I can't reproduce locally.

**Comment by Ivansete-status** [comment]
Just for our reference, the following PR is waiting for this #2761 one to be merged: https://github.com/status-im/status-go/pull/5123

Once merged we will need to ping @richard-ramos and then we will test this in `status.staging`

( cc @SionoiS )

**Comment by gabrielmer** [comment]
LGTM! Let's merge it once CI passes and after testing thoroughly - looks amazing 🔥

**Comment by Ivansete-status** [comment]
As mentioned in our recent Waku PM session, we will merge this once we make sure the following works well:
1. database migration
2. cursor issue works well from _Status_ perspective
3. tests performed by @stubbsta works well

File: migrations/message_store_postgres/content_script_version_6.nim
Comment (Ivansete-status) [inline migrations/message_store_postgres/content_script_version_6.nim:20]: @SionoiS - should we establish the PK as only the `MESSAGEHASH`?

Comment (richard-ramos) [reply migrations/message_store_postgres/content_script_version_6.nim:20]: Would it make sense to create an unique index per message hash to speed up store queries using message hash crtiteria?

**Comment by Ivansete-status** [comment]
Maybe we can simplify the migration script because we've been giving the `storedAt` column the `timestamp` values as can be seen from:

https://github.com/waku-org/nwaku/blob/dc337cd210d564d5fa1b8c4b2736e2b21baaafc5/waku/waku_archive/archive.nim#L116

https://github.com/waku-org/nwaku/blob/dc337cd210d564d5fa1b8c4b2736e2b21baaafc5/waku/waku_archive/driver/postgres_driver/postgres_driver.nim#L230

File: migrations/message_store_postgres/content_script_version_6.nim
Comment (Ivansete-status) [reply migrations/message_store_postgres/content_script_version_6.nim]: The initial comment that I did is outdated because we will follow a different migration approach

**Review by Ivansete-status — APPROVED** [review]

**Review by gabrielmer — APPROVED** [review]

**Comment by Ivansete-status** [comment]
Thank you so much guys for the patience and energy invested in this PR :raised_hands:
I think it looks nice! ( until we find another bug ;P )

**Comment by Ivansete-status** [comment]
I've tested that the schema migration works well in a 121GB database ( in wakudev host .)

**Pull request merged**
